release-25.3.0-rc: pgwire: limit concurrent updates to last_login_time

celiala opened the pull request:
Backport 4/4 commits from #150991, 1 commit from #151167, and part of 1 commit from #150562.

/cc @cockroachdb/release

Release justification: resolve a GA blocker issue by preventing excessive write load on a system table.

---

###  pgwire: limit concurrent updates to last_login_time

This patch uses singleflight to ensure that each node only has one
request in flight for updating the last_login_time. If there already is
a request in flight, then that user is added to a set of pending users
to update, and their time gets populated in the next singleflight.

This will reduce write traffic to the system.users table.

This includes a minor refactor to separate the login_time updating logic
into a separate component.

As part of this, we also now skip updating the last_login_time if running in
a read-only tenant.

###  sql/pgwire: use a single query to update last_login_time

Rather than using a separate query for each user, now we will run one
query to update multiple users at once.

###  sql/pgwire: avoid frequent updates to last_login_time

This patch adds a small LRU cache to avoid updating the
estimated_last_login_time overly frequently for a single user. The cache
will cause us to skip the update if the user's login time was last
updated within the past 10 minutes.

This requires us to update the tests to use a different user, since the
existing test drops `testuser` and recreates, which means that the
last_login_time will not be updated for that user as it's still in the
LRU cache.

----

fixes https://github.com/cockroachdb/cockroach/issues/151051
fixes https://github.com/cockroachdb/cockroach/issues/150560

Release note: None

rafiss committed:
pgwire: limit concurrent updates to last_login_time

This patch uses singleflight to ensure that each node only has one
request in flight for updating the last_login_time. If there already is
a request in flight, then that user is added to a set of pending users
to update, and their time gets populated in the next singleflight.

This will reduce write traffic to the system.users table.

This includes a minor refactor to separate the login_time updating logic
into a separate component.

Release note: None

rafiss committed:
sql/pgwire: use a single query to update last_login_time

Rather than using a separate query for each user, now we will run one
query to update multiple users at once.

Release note: None

rafiss committed:
sql/pgwire: avoid frequent updates to last_login_time

This patch adds a small LRU cache to avoid updating the
estimated_last_login_time overly frequently for a single user. The cache
will cause us to skip the update if the user's login time was last
updated within the past 10 minutes.

This requires us to update the tests to use a different user, since the
existing test drops `testuser` and recreates, which means that the
last_login_time will not be updated for that user as it's still in the
LRU cache.

Release note: None

rafiss committed:
pgwire: use defer to unlock lastLoginUpdater mutex

This makes the code resilient to panics.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151150)
<!-- Reviewable:end -->

rafiss committed:
server: add TenantReadOnly flag to ExecutorConfig

Add TenantReadOnly field to SQLConfig and ExecutorConfig to track
tenant-level read-only status. Modify tenant initialization to detect
read-only tenants by checking mtinfopb.TenantInfoWithUsage.ReadFromTenant.

Cherry-picked from ab73c65f119d98e44e8716fe1f134a8eb7c49370
(excluding stats-related files)

Release note: None

yuzefovich commented:
Heads up that there appears to be a data race in this change - see #151164.

rafiss committed:
pgwire: protect lastUpdateCache with a mutex

The cache is not safe for concurrent access so it needs to be protected
with the mutex.

Release note: None
